src/internal/connector/graph/service.go

```diff
@@ -319,3 +365,181 @@ func RunWithRetry(run func() error) error {
 
 	return support.ConnectorStackErrorTraceWrap(err, "maximum retries or unretryable")
 }
+
+// RetryHandler handles transient HTTP responses and retries the request given the retry options
+type RetryHandler struct {
+	// default options to use when evaluating the response
+	options RetryHandlerOptions
+}
+
+type RetryHandlerOptions struct {
+	// Callback to determine if the request should be retried
+	ShouldRetry func(delay time.Duration, executionCount int, request *http.Request, response *http.Response) bool
+	// The maximum number of times a request can be retried
+	MaxRetries int
+	// The delay in seconds between retries
+	DelaySeconds int
+}
+
+type retryHandlerOptionsInt interface {
+	abs.RequestOption
+	GetShouldRetry() func(delay time.Duration, executionCount int,
+		request *http.Request, response *http.Response) bool
+	GetDelaySeconds() int
+	GetMaxRetries() int
+}
+
+var retryKeyValue = abs.RequestOptionKey{
+	Key: "RetryHandler",
+}
+
+// GetKey returns the key value to be used when the option is added to the request context
+func (options *RetryHandlerOptions) GetKey() abs.RequestOptionKey {
+	return retryKeyValue
+}
+
+// GetShouldRetry returns the should retry callback function which evaluates the response for retrial
+func (options *RetryHandlerOptions) GetShouldRetry() func(delay time.Duration,
+	executionCount int, request *http.Request, response *http.Response) bool {
+	return options.ShouldRetry
+}
+
+// GetDelaySeconds returns the delays in seconds between retries
+func (options *RetryHandlerOptions) GetDelaySeconds() int {
+	if options.DelaySeconds < 1 {
+		return defaultDelaySeconds
+	}
+
+	return options.DelaySeconds
+}
+
+// GetMaxRetries returns the maximum number of times a request can be retried
+func (options *RetryHandlerOptions) GetMaxRetries() int {
+	if options.MaxRetries < 1 {
+		return defaultMaxRetries
+	}
+
+	return options.MaxRetries
+}
+
+// Intercept implements the interface and evaluates whether to retry a failed request.
+func (middleware RetryHandler) Intercept(
+	pipeline khttp.Pipeline,
+	middlewareIndex int,
+	req *http.Request,
+) (*http.Response, error) {
+	obsOptions := khttp.GetObservabilityOptionsFromRequest(req)
+	ctx := req.Context()
+
+	var (
+		span              trace.Span
+		observabilityName string
+	)
+
+	if obsOptions != nil {
+		observabilityName = obsOptions.GetTracerInstrumentationName()
+		ctx, span = otel.GetTracerProvider().Tracer(observabilityName).Start(ctx, "RetryHandler_Intercept")
+		span.SetAttributes(attribute.Bool("com.microsoft.kiota.handler.retry.enable", true))
+
+		defer span.End()
+
+		req = req.WithContext(ctx)
+	}
+
+	response, err := pipeline.Next(req, middlewareIndex)
+	if err != nil {
+		return response, err
+	}
+
+	reqOption, ok := req.Context().Value(retryKeyValue).(retryHandlerOptionsInt)
+	if !ok {
+		reqOption = &middleware.options
+	}
+
+	return middleware.retryRequest(ctx, pipeline,
+		middlewareIndex, reqOption, req, response, 0, 0, observabilityName)
```

Comment:
Nit: code standards, all params should be inline, or each on their own line (similar types can be stacked).
```suggestion
	return middleware.retryRequest(
	  ctx,
	  pipeline,
	  middlewareIndex,
	  reqOption,
	  req,
	  response,
	  0, 0,
	  observabilityName)
```